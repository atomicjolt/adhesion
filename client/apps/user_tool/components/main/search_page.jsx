import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { searchForAccountUsers } from '../../actions/application';
import UserSearchResult from './user_search_result';

const select = state => ({
  matchingUsers: state.application.matchingUsers,
  lmsAccountId: state.settings.custom_canvas_account_id,
});

export class SearchPage extends React.Component {
  static propTypes = {
    searchForAccountUsers: PropTypes.func.isRequired,
    matchingUsers: PropTypes.array.isRequired,
    lmsAccountId: PropTypes.string.isRequired,
  };

  constructor() {
    super();

    this.state = { searchTerm: '' };
    this.minSearchTermLength = 3;
  }

  updateSearchTerm(event) {
    this.setState({ searchTerm: event.target.value });
  }

  handleSearch(event) {
    event.preventDefault();
    event.target.form.reportValidity();

    const { lmsAccountId, searchForAccountUsers:search } = this.props;
    const { searchTerm } = this.state;

    if (searchTerm.length >= this.minSearchTermLength) {
      search(lmsAccountId, searchTerm);
    }
  }

  render() {
    const { matchingUsers } = this.props;
    const { searchTerm } = this.state;
    const renderedUsers = matchingUsers.map(user => (
      <UserSearchResult key={user.id} user={user} />
    ));

    return (
      <div>
        <form>
          <input
            type="search"
            minLength={this.minSearchTermLength}
            value={searchTerm}
            onChange={event => this.updateSearchTerm(event)}
            placeholder="Search for students..."
          />
          <button type="submit" onClick={event => this.handleSearch(event)}>Search</button>
        </form>
        <p>Search Results:</p>
        <table>
          <thead>
            <tr>
              <th scope="col">Name</th>
              <th scope="col">Email Address</th>
              <th scope="col">Roles</th>
              <th scope="col">Login ID</th>
              <th scope="col">SIS ID</th>
            </tr>
          </thead>
          <tbody>
            {renderedUsers}
          </tbody>
        </table>
      </div>
    );
  }
}

export default connect(select, { searchForAccountUsers })(SearchPage);
