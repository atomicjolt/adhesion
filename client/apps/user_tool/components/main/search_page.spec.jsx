import React from 'react';
import { shallow, mount } from 'enzyme';

import { SearchPage } from './search_page';

jest.mock('./edit_user_modal');

describe('SearchPage', () => {
  const props = {
    searchForAccountUsers: () => {},
    matchingUsers: [
      {
        id: 1,
        name: 'George Washington',
        login_id: 'countryfather@revolution.com',
        sis_user_id: 'george_123',
        roles: ['admin', 'teacher'],
        email: 'countryfather@revolution.com',
      },
      {
        id: 2,
        name: 'Thomas Jefferson',
        login_id: 'idodeclare@revolution.com',
        sis_user_id: 'thomas_123',
        roles: ['teacher'],
        email: 'idodeclare@revolution.com',
      }
    ],
    lmsAccountId: '1',
    currentPage: 2,
    previousPageAvailable: true,
    nextPageAvailable: true,
  };

  const submitSearch = (searchPage) => {
    searchPage.find('button[type="submit"]').simulate('click', {
      preventDefault: () => {},
      target: { form: { reportValidity: () => {} } },
    });
  };

  it('renders the search page', () => {
    const searchPage = shallow(<SearchPage
      matchingUsers={props.matchingUsers}
      searchForAccountUsers={props.searchForAccountUsers}
      lmsAccountId={props.lmsAccountId}
      currentPage={props.currentPage}
      previousPageAvailable={props.previousPageAvailable}
      nextPageAvailable={props.nextPageAvailable}
    />);

    expect(searchPage).toMatchSnapshot();
  });

  describe('when the user submits a search', () => {
    it('submits a search request', () => {
      spyOn(props, 'searchForAccountUsers');
      const searchTerm = 'student name';
      const searchPage = shallow(<SearchPage
        matchingUsers={props.matchingUsers}
        searchForAccountUsers={props.searchForAccountUsers}
        lmsAccountId={props.lmsAccountId}
        currentPage={props.currentPage}
      />);

      searchPage.find('input').simulate('change', { target: { value: searchTerm } });
      submitSearch(searchPage);

      expect(props.searchForAccountUsers).toHaveBeenCalledWith(
        props.lmsAccountId,
        searchTerm,
      );
    });
  });

  describe('when the user submits a search with less than 3 characters', () => {
    it('does not submit the search', () => {
      spyOn(props, 'searchForAccountUsers');
      const searchTerm = 'jo';
      const searchPage = shallow(<SearchPage
        matchingUsers={props.matchingUsers}
        searchForAccountUsers={props.searchForAccountUsers}
        lmsAccountId={props.lmsAccountId}
        currentPage={props.currentPage}
      />);

      searchPage.find('input').simulate('change', { target: { value: searchTerm } });
      submitSearch(searchPage);

      expect(props.searchForAccountUsers).not.toHaveBeenCalled();
    });

    it('does not affect the paged results', () => {
      spyOn(props, 'searchForAccountUsers');
      const firstSearchTerm = 'john';
      const shortSearchTerm = 'sa';
      const searchPage = mount(<SearchPage
        matchingUsers={props.matchingUsers}
        searchForAccountUsers={props.searchForAccountUsers}
        lmsAccountId={props.lmsAccountId}
        currentPage={props.currentPage}
        previousPageAvailable={props.previousPageAvailable}
        nextPageAvailable={props.nextPageAvailable}
      />);

      searchPage.find('input').simulate('change', { target: { value: firstSearchTerm } });
      submitSearch(searchPage);

      expect(props.searchForAccountUsers).toHaveBeenCalledWith(
        props.lmsAccountId,
        firstSearchTerm,
      );

      searchPage.find('input').simulate('change', { target: { value: shortSearchTerm } });
      submitSearch(searchPage);

      const buttons = searchPage.find('button');
      const nextButton = buttons.at(buttons.length - 1);
      nextButton.simulate('click');

      expect(props.searchForAccountUsers).toHaveBeenCalledWith(
        props.lmsAccountId,
        firstSearchTerm,
        props.currentPage + 1,
      );
    });
  });

  describe('when the user updates the value in the search input field', () => {
    it('does not affect the paged results', () => {
      spyOn(props, 'searchForAccountUsers');
      const firstSearchTerm = 'john';
      const secondSearchTerm = 'jones';
      const searchPage = mount(<SearchPage
        matchingUsers={props.matchingUsers}
        searchForAccountUsers={props.searchForAccountUsers}
        lmsAccountId={props.lmsAccountId}
        currentPage={props.currentPage}
        previousPageAvailable={props.previousPageAvailable}
        nextPageAvailable={props.nextPageAvailable}
      />);

      searchPage.find('input').simulate('change', { target: { value: firstSearchTerm } });
      submitSearch(searchPage);

      expect(props.searchForAccountUsers).toHaveBeenCalledWith(
        props.lmsAccountId,
        firstSearchTerm,
      );

      searchPage.find('input').simulate('change', { target: { value: secondSearchTerm } });

      const buttons = searchPage.find('button');
      const nextButton = buttons.at(buttons.length - 1);
      nextButton.simulate('click');

      expect(props.searchForAccountUsers).toHaveBeenCalledWith(
        props.lmsAccountId,
        firstSearchTerm,
        props.currentPage + 1,
      );
    });
  });
});
