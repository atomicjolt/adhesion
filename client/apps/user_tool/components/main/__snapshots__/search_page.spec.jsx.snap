// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage renders the search page 1`] = `
<div
  className="main"
>
  <form
    className="search"
    role="search"
  >
    <label
      className="hidden"
      htmlFor="search"
    >
      Search
    </label>
    <input
      id="search"
      minLength={3}
      name="search"
      onChange={[Function]}
      placeholder="Search for students..."
      type="search"
      value=""
    />
    <button
      className="search__btn"
      onClick={[Function]}
      type="submit"
    >
      Search
    </button>
  </form>
  <div
    className="table-scroll"
  >
    <table
      className="search-results"
    >
      <caption
        className="hidden"
      >
        Search Results
      </caption>
      <thead>
        <tr>
          <th
            scope="col"
          >
            Name
          </th>
          <th
            scope="col"
          >
            Login ID
          </th>
          <th
            scope="col"
          >
            SIS ID
          </th>
          <th
            scope="col"
          >
            Roles
          </th>
          <th
            scope="col"
          >
            Email Address
          </th>
        </tr>
      </thead>
      <tbody>
        <UserSearchResult
          key="1"
          user={
            Object {
              "email": "countryfather@revolution.com",
              "id": 1,
              "login_id": "countryfather@revolution.com",
              "name": "George Washington",
              "roles": Array [
                "admin",
                "teacher",
              ],
              "sis_user_id": "george_123",
            }
          }
        />
        <UserSearchResult
          key="2"
          user={
            Object {
              "email": "idodeclare@revolution.com",
              "id": 2,
              "login_id": "idodeclare@revolution.com",
              "name": "Thomas Jefferson",
              "roles": Array [
                "teacher",
              ],
              "sis_user_id": "thomas_123",
            }
          }
        />
      </tbody>
    </table>
  </div>
  <StartSearching />
  <Pagination
    changePageTo={[Function]}
    currentPage={2}
    nextPageAvailable={true}
    previousPageAvailable={true}
  />
</div>
`;

exports[`SearchPage when the user submits a search that returns no results displays the no search results section 1`] = `
<div
  className="main"
>
  <form
    className="search"
    role="search"
  >
    <label
      className="hidden"
      htmlFor="search"
    >
      Search
    </label>
    <input
      id="search"
      minLength={3}
      name="search"
      onChange={[Function]}
      placeholder="Search for students..."
      type="search"
      value="student name"
    />
    <button
      className="search__btn"
      onClick={[Function]}
      type="submit"
    >
      Search
    </button>
  </form>
  <div
    className="table-scroll"
  >
    <table
      className="search-results"
    >
      <caption
        className="hidden"
      >
        Search Results
      </caption>
      <thead>
        <tr>
          <th
            scope="col"
          >
            Name
          </th>
          <th
            scope="col"
          >
            Login ID
          </th>
          <th
            scope="col"
          >
            SIS ID
          </th>
          <th
            scope="col"
          >
            Roles
          </th>
          <th
            scope="col"
          >
            Email Address
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
  <NoSearchResults
    searchTerm="student name"
  />
  <Pagination
    changePageTo={[Function]}
    currentPage={2}
    nextPageAvailable={true}
    previousPageAvailable={true}
  />
</div>
`;
